### PR TITLE
UI for creating API keys

### PR DESCRIPTION
**Pull request by bion**
- [x] admin can fill out and submit a form to create a new API key
- [x] form shows validation errors when needed
- [x] credential viewing page after key is created
- [x] browser test for happy path
- [x] handle ProgramNotFoundException with new client-error handler

part of https://github.com/seattle-uat/civiform/issues/1744

Screenshot from 2022-05-10 09-18-21
Screenshot from 2022-05-10 10-06-15

### CODE REVIEW

File: server/app/views/admin/apikeys/ApiKeyCredentialsView.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    ContainerTag contentDiv =
+        div()
+            .withClasses(Styles.PX_20)
+            .with(
+                h1(title).withClasses(Styles.MY_4),
+                h2("Credentials"),
+                p(
+                    "Please copy your API key and it store it somewhere secure. This is your only"
+                        + " opportunity to copy the secret from CiviForm, if you "
+                        + "refresh the page or navigate away you will not be able to recover the"
```
Comment (shanemc-goog) [inline server/app/views/admin/apikeys/ApiKeyCredentialsView.java:41]: Consistently have the continuation space character at the start or end of the different segments.

File: server/app/views/admin/apikeys/ApiKeyCredentialsView.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    String title = "Created API key: " + apiKey.getName();
+
+    ContainerTag contentDiv =
+        div()
+            .withClasses(Styles.PX_20)
+            .with(
+                h1(title).withClasses(Styles.MY_4),
+                h2("Credentials"),
+                p(
+                    "Please copy your API key and it store it somewhere secure. This is your only"
```
Comment (shanemc-goog) [inline server/app/views/admin/apikeys/ApiKeyCredentialsView.java:39]: suggestion: It's probably cleaner to have this text as a class constant so that it doesn't need to wrap so much or divide up the page construction.

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+                        FieldWithLabel.input()
+                            .setFieldName(ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
+                            .setId(ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
+                            .setLabelText("API key name"),
+                        dynamicForm,
+                        ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
+                    .getContainer(),
+                h2("Expiration date"),
+                p(
+                    "Specify a date when this API key will no longer be valid. The expiration date"
```
Comment (shanemc-goog) [inline server/app/views/admin/apikeys/ApiKeyNewOneView.java:66]: likewise to the other html page, making this a const in the file will likely aid readability all around.

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                        + " allows only the IP 8.8.8.8 to use the API key. All IP addresses can be"
+                        + " allowed, but this is strongly discouraged for security reasons because"
+                        + " it would allow any computer connected to the internet to use the key"
+                        + " if they obtain it. All IP addresses can be allowed with a value of"
+                        + " 0.0.0.0/0."),
+                setStateIfPresent(
+                        FieldWithLabel.input()
+                            .setFieldName(ApiKeyService.FORM_FIELD_NAME_SUBNET)
+                            .setId(ApiKeyService.FORM_FIELD_NAME_SUBNET)
+                            .setLabelText("API key name"),
```
Comment (shanemc-goog) [inline server/app/views/admin/apikeys/ApiKeyNewOneView.java]: I think "key name" needs updating for this field?  I get lost in the page construction so I might be off.

File: server/app/controllers/admin/AdminApiKeysController.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    try {
+      result = apiKeyService.createApiKey(form, profile.get());
+    } catch (ProgramNotFoundException e) {
+      return notFound(e.toString());
+    }
+
+    if (!result.isSuccessful()) {
+      return badRequest(
+          newOneView.render(
+              request, programService.getAllProgramNames(), Optional.of(result.getForm())));
```
Comment (shanemc-goog) [inline server/app/controllers/admin/AdminApiKeysController.java:79]: I don't believe getAllProgramsNames filters for program version, meaning it will show a Draft I think.  Probably best to explicitly get Active programs.

**Review by shanemc-goog — CHANGES_REQUESTED** [review]

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java (hunk 4 above)
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java]: done

File: server/app/views/admin/apikeys/ApiKeyCredentialsView.java (hunk 2 above)
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyCredentialsView.java:39]: done

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java (hunk 3 above)
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:66]: done

File: server/app/views/admin/apikeys/ApiKeyCredentialsView.java (hunk 1 above)
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyCredentialsView.java:41]: done

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+                    .getContainer(),
+                h2("Allowed IP addresses"),
+                p(
+                    "Specify a subnet using CIDR notation that is allowed to call the API. The"
+                        + " subnet may be edited after the API key is created. A single IP address"
+                        + " can be specified with a mask of /32. For example, \"8.8.8.8/32\""
+                        + " allows only the IP 8.8.8.8 to use the API key. All IP addresses can be"
+                        + " allowed, but this is strongly discouraged for security reasons because"
+                        + " it would allow any computer connected to the internet to use the key"
+                        + " if they obtain it. All IP addresses can be allowed with a value of"
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:88]: done

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+                        FieldWithLabel.date()
+                            .setFieldName(ApiKeyService.FORM_FIELD_NAME_EXPIRATION)
+                            .setId(ApiKeyService.FORM_FIELD_NAME_EXPIRATION)
+                            .setLabelText("Expiration date"),
+                        dynamicForm,
+                        ApiKeyService.FORM_FIELD_NAME_EXPIRATION)
+                    .getContainer(),
+                h2("Allowed IP addresses"),
+                p(
+                    "Specify a subnet using CIDR notation that is allowed to call the API. The"
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:82]: done

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+                            .setFieldName(ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
+                            .setId(ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
+                            .setLabelText("API key name"),
+                        dynamicForm,
+                        ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
+                    .getContainer(),
+                h2("Expiration date"),
+                p(
+                    "Specify a date when this API key will no longer be valid. The expiration date"
+                        + " may be edited after the API is created. It is strongly recommended to"
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:67]: done

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    String title = "Create a new API key";
+
+    ContainerTag formTag =
+        form()
+            .withMethod("POST")
+            .with(
+                makeCsrfTokenInputTag(request),
+                setStateIfPresent(
+                        FieldWithLabel.input()
+                            .setFieldName(ApiKeyService.FORM_FIELD_NAME_KEY_NAME)
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:58]: made it more consistent

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (shanemc-goog) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:82]: don't see a change for this or the other ones you commented on, did you push the commit?

File: browser-test/src/support/index.ts [hunk 10]
```diff
@@ -8,11 +9,13 @@ export { AdminTIGroups } from './admin_ti_groups'
 export { ApplicantQuestions } from './applicant_questions'
 export { clickAndWaitForModal, waitForPageJsLoad } from './wait'
 
-const {
-  BASE_URL = 'http://civiform:9000',
-  TEST_USER_LOGIN = '',
-  TEST_USER_PASSWORD = '',
-} = process.env
+import {
```
Comment (shanemc-goog) [inline browser-test/src/support/index.ts:12]: just curious: why this change?

File: server/app/annotations/BindingAnnotations.java [hunk 11]
```diff
@@ -18,4 +19,12 @@ public final class BindingAnnotations {
...
   @Retention(RUNTIME)
   public @interface Now {}
+
+  /**
+   * Holds the {@link play.i18n.Messages} object associated with the "en-US" {@link play.i18n.Lang}.
+   */
+  @Qualifier
+  @Target({FIELD, PARAMETER, METHOD})
+  @Retention(RUNTIME)
+  public @interface EnUs {}
```
Comment (shanemc-goog) [inline server/app/annotations/BindingAnnotations.java:29]: Suggestion: qualify this as a language.  EnUsLang maybe.

EnUs is on the edge of looking random if you just come across it, and if say we did portuguese instead @Pt would be incredibly unreadable on a method.

File: server/app/controllers/admin/AdminApiKeysController.java [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+    if (!profile.isPresent()) {
+      throw new RuntimeException("Unable to resolve profile.");
+    }
+
+    DynamicForm form = formFactory.form().bindFromRequest(request);
+
+    ApiKeyCreationResult result;
+    try {
+      result = apiKeyService.createApiKey(form, profile.get());
+    } catch (ProgramNotFoundException e) {
```
Comment (shanemc-goog) [inline server/app/controllers/admin/AdminApiKeysController.java:72]: unrelated: but re the ErrorHandler discussion, this could be handled there and converted to a RuntimeException.

File: server/app/controllers/admin/AdminApiKeysController.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+      return notFound(e.toString());
+    }
+
+    if (!result.isSuccessful()) {
```
Comment (shanemc-goog) [inline server/app/controllers/admin/AdminApiKeysController.java:76]: suggestion: Flip the conditions, positive assertions are generally more readable.

File: server/conf/routes [hunk 14]
```diff
@@ -85,6 +85,11 @@ GET     /admin/programs/:programId/applications/:applicationId            contro
 GET     /admin/programs/:programId/applications/:applicationId/download   controllers.admin.AdminApplicationController.download(request: Request, programId: Long, applicationId: Long)
 GET     /admin/demographics                                               controllers.admin.AdminApplicationController.downloadDemographics()
 
+# Admin API key management
+GET     /admin/apiKeys                              controllers.admin.AdminApiKeysController.index()
+GET     /admin/apiKeys/new                          controllers.admin.AdminApiKeysController.newOne(request: Request)
```
Comment (shanemc-goog) [inline server/conf/routes:90]: question: newOne and create don't read too differently from me, is it normal in REST to not have the phrasing of them be more indicative of what you're getting? or does Get/Post largely make that distinction.  EG a GET wouldn't have side effects so newOne can be inferred to be a page that facilitates making a new one.

File: server/app/controllers/admin/AdminApiKeysController.java (hunk 5 above)
Comment (bion) [reply server/app/controllers/admin/AdminApiKeysController.java:79]: done

File: browser-test/src/support/index.ts (hunk 10 above)
Comment (bion) [reply browser-test/src/support/index.ts:12]: so that BASE_URL could be use by other files in `src/support` without creating a circular dependency

File: server/app/controllers/admin/AdminApiKeysController.java (hunk 13 above)
Comment (bion) [reply server/app/controllers/admin/AdminApiKeysController.java:76]: done

File: server/app/annotations/BindingAnnotations.java (hunk 11 above)
Comment (bion) [reply server/app/annotations/BindingAnnotations.java]: done

File: server/app/controllers/admin/AdminApiKeysController.java (hunk 12 above)
Comment (bion) [reply server/app/controllers/admin/AdminApiKeysController.java:72]: yessss nice i'm gonna do it

Comment (bion) [reply server/app/controllers/admin/AdminApiKeysController.java]: done

File: server/app/views/admin/apikeys/ApiKeyCredentialsView.java [hunk 15]
```diff
@@ -21,6 +21,12 @@
 public final class ApiKeyCredentialsView extends BaseHtmlView {
   private final AdminLayout layout;
 
+  private static final String CREDENTIALS_DESCRIPTION =
+      "Please copy your API key and it store it somewhere secure. This is your only"
+          + " opportunity to copy the secret from CiviForm, if you "
+          + "refresh the page or navigate away you will not be able to recover the"
```
Comment (shanemc-goog) [inline server/app/views/admin/apikeys/ApiKeyCredentialsView.java:27]: Move the trailing space before this to a leading space here.

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java [hunk 16]
```diff
@@ -32,8 +35,33 @@ public final class ApiKeyNewOneView extends BaseHtmlView {
...
+          + " recommended. Once a year is acceptable in most situations.";
+  private static final DomContent[] SUBNET_DESCRIPTION = {
+    text("Specify a subnet using "),
+    new LinkElement()
+        .setText("CIDR notation")
+        .setHref("https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing")
+        .opensInNewTab()
+        .asAnchorText(),
+    text(
+        " that is allowed to call the API. The"
```
Comment (shanemc-goog) [inline server/app/views/admin/apikeys/ApiKeyNewOneView.java:53]: I think some of the text lines were deleted accidentally

File: server/app/services/apikey/ApiKeyService.java [hunk 17]
```diff
@@ -206,7 +205,10 @@ private ApiKeyGrants resolveGrants(DynamicForm form) throws ProgramNotFoundExcep
       String programSlug = matcher.group(1);
 
       if (!programSlugs.contains(programSlug)) {
-        throw new ProgramNotFoundException(programSlug);
+        // Making the root exception cause a ProgramNotFoundException will cause
+        // controllers.ErrorHandler to handle it
+        // as a client-error, rather than server-error.
+        throw new RuntimeException(new ProgramNotFoundException(programSlug));
```
Comment (shanemc-goog) [inline server/app/services/apikey/ApiKeyService.java:211]: discussion: This is fine, though subsequently, I think it'd be good/fine to make ProgramNotFoundException a Runtime Exception.  If code really needs to know it can do a hasProgramId() kinda check, but I'm guessing most just catch and wrap this as notFound without doing any other recovery or reporting.

**Review by shanemc-goog — CHANGES_REQUESTED** [review]

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java (hunk 16 above)
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:53]: looks right to me?

File: server/app/services/apikey/ApiKeyService.java (hunk 17 above)
Comment (bion) [reply server/app/services/apikey/ApiKeyService.java:211]: Yeah, to make the change we'll need to take a look at all the places `ProgramNotFoundException` is being used and see if any of them are using it for control flow that's fundamentally different from what the default behavior with the new pattern does. If so we should probably leave it as a checked exception.

File: server/app/views/admin/apikeys/ApiKeyCredentialsView.java (hunk 15 above)
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyCredentialsView.java]: done

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java (hunk 16 above)
Comment (shanemc-goog) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:53]: yeah, I'd deleted. I didn't see the construction just above.  But GH is duplicating my comments today.... :/

**Review by shanemc-goog — APPROVED** [review]

File: server/app/controllers/ErrorHandler.java [hunk 18]
```diff
@@ -55,11 +60,12 @@ public CompletionStage<Result> onServerError(RequestHeader request, Throwable ex
...
    * our application exception. Anecdotally it's 2 levels down.
    */
-  static Optional<Throwable> findThrowableByType(Throwable exception, Class<?> search) {
+  static Optional<Throwable> findThrowableByTypes(
+      Throwable exception, ImmutableSet<Class<?>> search) {
     Optional<Throwable> root = Optional.of(exception);
     // Search a couple causes deep for the desired type.
     for (int i = 0; i < 5 && root.isPresent(); ++i) {
-      if (search.isInstance(root.get())) {
+      if (search.contains(root.get().getClass())) {
```
Comment (clouser2) [inline server/app/controllers/ErrorHandler.java:68]: Not-blocking: Previously, we'd match subclasses of the provided "search" parameter, whereas now it has to be an exact match.

I'd suggest documenting that and marking the two specific error types that we currently handle as "final" to prevent us trying to take any implicit dependency on that behavior.

File: server/app/controllers/admin/AdminApiKeysController.java [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+    this.apiKeyService = checkNotNull(apiKeyService);
+    this.newOneView = checkNotNull(newOneView);
+    this.apiKeyCredentialsView = checkNotNull(apiKeyCredentialsView);
+    this.programService = checkNotNull(programService);
+    this.formFactory = checkNotNull(formFactory);
+    this.profileUtils = checkNotNull(profileUtils);
+  }
+
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
+  public Result index() {
```
Comment (clouser2) [inline server/app/controllers/admin/AdminApiKeysController.java:49]: Not-blocking: Consider removing this and reintroducing it when it's implemented. Alternatively, leave a TODO? Or ignore entirely, up to you.

File: server/app/controllers/admin/AdminApiKeysController.java [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+    return ok();
+  }
+
+  @Secure(authorizers = Authorizers.Labels.CIVIFORM_ADMIN)
```
Comment (clouser2) [inline server/app/controllers/admin/AdminApiKeysController.java:53]: Not-blocking: Out of curiosity, if all handlers in a controller require the same authorizer, is it possible to annotate at the class level instead? This would help prevent the case where a new handler is added to an existing controller and forgot to add an authorizer. It would potentially also reduce code duplication.

File: server/app/controllers/admin/AdminApiKeysController.java [hunk 21]
```diff
@@ -0,0 +1,77 @@
...
+    Optional<CiviFormProfile> profile = profileUtils.currentUserProfile(request);
+
+    if (!profile.isPresent()) {
+      throw new RuntimeException("Unable to resolve profile.");
+    }
+
+    DynamicForm form = formFactory.form().bindFromRequest(request);
+    ApiKeyCreationResult result = apiKeyService.createApiKey(form, profile.get());
+
+    if (result.isSuccessful()) {
```
Comment (clouser2) [inline server/app/controllers/admin/AdminApiKeysController.java:69]: Not-blocking: Complete style nit. Consider putting the error handling in the if-block and putting the "happy" case as the last statement. This is mostly a style preference that reading through the code ends in the case we expect. A perhaps weaker argument is that there's another indented block in this code which throws an exception, so it's perhaps slightly less readable to have the next indented block actually return the success case. My preference is pretty weak here, so feel free to discard if you'd like.

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+            + " 0.0.0.0/0, though this is not recommended.")
+  };
+
+  @Inject
+  public ApiKeyNewOneView(AdminLayout layout, @EnUsLang Messages enUsMessages) {
+    this.layout = checkNotNull(layout);
+    this.enUsMessages = checkNotNull(enUsMessages);
+  }
+
+  public Content render(Request request, ImmutableSet<String> programNames) {
```
Comment (clouser2) [inline server/app/views/admin/apikeys/ApiKeyNewOneView.java:69]: Not-blocking: Since there's only one caller of this, consider removing the method overload and having the controller call render(request, programNames, Optional.empty()).

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+
+    formTag.with(h2("Allowed programs"), p("Select the programs this key grants read access to."));
+
+    for (String name : programNames.stream().sorted().collect(ImmutableList.toImmutableList())) {
+      formTag.with(
+          FieldWithLabel.checkbox()
+              .setFieldName(programReadGrantFieldName(name))
+              .setLabelText(name)
+              .setId(slugifier.slugify(name))
+              .setValue("true")
```
Comment (clouser2) [inline server/app/views/admin/apikeys/ApiKeyNewOneView.java:121]: Does this imply that new keys grant access to all current programs by default? I'd consider not selecting anything by default and having a "select all" option. My rationale for this is that it aligns with the principle-of-least-privilege.

This point / distinction needn't block the larger review at all though.

File: server/app/views/components/FieldWithLabel.java [hunk 24]
```diff
@@ -223,7 +224,20 @@ public FieldWithLabel setScreenReaderText(String screenReaderText) {
...
     this.messages = messages;
-    this.fieldErrors = errors;
+    this.fieldErrors =
+        errors.stream()
+            .map(
+                (ValidationErrorMessage vem) ->
+                    new ValidationError(vem.key().getKeyName(), vem.key().getKeyName(), vem.args()))
+            .collect(ImmutableSet.toImmutableSet());
+
+    return this;
```
Comment (clouser2) [inline server/app/views/components/FieldWithLabel.java:234]: Not-blocking: Nit. Since this method overload will hopefully remain temporary, what about having it call the more preferred method with the converted errors.

e.g.

public FieldWithLabel setFieldErrors(Messages messages, ImmutableSet<ValidationErrorMessage> errors) {
  return setFieldErrors(messages, errors.stream().....);
}

**Review by clouser2 — APPROVED** [review]

File: server/app/controllers/admin/AdminApiKeysController.java (hunk 20 above)
Comment (bion) [reply server/app/controllers/admin/AdminApiKeysController.java:53]: The docs don't seem to suggest that you can: https://github.com/pac4j/play-pac4j/wiki/Apply-security

I tried moving the annotation to the class and while the app did run, I got some strange behavior when trying to access the secured routes as an authorized user.

File: server/app/controllers/admin/AdminApiKeysController.java (hunk 21 above)
Comment (bion) [reply server/app/controllers/admin/AdminApiKeysController.java:69]: hahah that was actually its original structure, @shanemc-goog requested a change to avoid having `if (!`. i don't have a strong opinion either.

File: server/app/controllers/admin/AdminApiKeysController.java (hunk 19 above)
Comment (bion) [reply server/app/controllers/admin/AdminApiKeysController.java:49]: it's already implemented in the next PR, i'll leave it as-is for now if that's alright

File: browser-test/src/admin_api_keys.test.ts [hunk 25]
```diff
@@ -0,0 +1,30 @@
...
+  it('Creating a new API key displays its credentials', async () => {
+    const { browser, page } = await startSession()
+
+    await loginAsAdmin(page)
+    const adminApiKeys = new AdminApiKeys(page)
+    const adminPrograms = new AdminPrograms(page)
+
+    const programName = 'API using program'
+    const programDescription = 'This program uses the API.'
+    await adminPrograms.addProgram(programName, programDescription, '', false)
```
[the remark this answers is not in the document]
Comment (bion) [reply browser-test/src/admin_api_keys.test.ts:18]: agree that's a good change, going to hold off as it's out of scope, as you mention

File: browser-test/src/admin_api_keys.test.ts [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+    await adminPrograms.publishAllPrograms()
+
+    const credentials = await adminApiKeys.createApiKey(
```
[the remark this answers is not in the document]
Comment (bion) [reply browser-test/src/admin_api_keys.test.ts:21]: done

File: browser-test/src/admin_api_keys.test.ts [hunk 27]
```diff
@@ -0,0 +1,30 @@
...
+      'Test API key',
+      '2022-01-31',
+      "8.8.8.8/32",
+      ['api-using-program']
+    )
+
+    expect(typeof credentials).toEqual("string")
```
[the remark this answers is not in the document]
Comment (bion) [reply browser-test/src/admin_api_keys.test.ts:28]: yeah, planning to do this once there's an API endpoint to call

File: server/app/controllers/ErrorHandler.java [hunk 28]
```diff
@@ -40,7 +42,10 @@ public ErrorHandler(
   public CompletionStage<Result> onServerError(RequestHeader request, Throwable exception) {
     // Exceptions that reach here will generate 500s. Here we convert certain ones to different user
     // visible states. Note: there are methods on the parent that handle dev and prod separately.
-    Optional<Throwable> match = findThrowableByType(exception, NotChangeableException.class);
+    Optional<Throwable> match =
+        findThrowableByTypes(
+            exception,
+            ImmutableSet.of(NotChangeableException.class, ProgramNotFoundException.class));
```
[the remark this answers is not in the document]
Comment (bion) [reply server/app/controllers/ErrorHandler.java:48]: done

File: server/app/controllers/ErrorHandler.java (hunk 18 above)
Comment (bion) [reply server/app/controllers/ErrorHandler.java:68]: good call! done

File: server/app/controllers/admin/AdminApiKeysController.java (hunk 21 above)
Comment (shanemc-goog) [reply server/app/controllers/admin/AdminApiKeysController.java:69]: hah lol yeah.  Happy either way, I can see brett's case too.

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java (hunk 23 above)
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:121]: The opposite -- the browser only sends form data for checkbox inputs that are checked. So the default is to grant access to no programs.

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java (hunk 22 above)
Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:69]: There are two callers: the `newOne` method for the initial render, and the `create` method when a submission fails validation -- both `render` method versions get used.

File: server/app/views/components/FieldWithLabel.java (hunk 24 above)
Comment (bion) [reply server/app/views/components/FieldWithLabel.java:234]: Unfortunately `FieldWithLabel setFieldErrors(Messages messages, ImmutableSet<ValidationError> errors)` and `setFieldErrors(Messages messages, ImmutableSet<ValidationErrorMessage> errors)` have the same erasure so we can't do that overload. It is temporary though since we'll do the broader refactor after this PR goes in.

File: browser-test/src/admin_api_keys.test.ts (hunk 27 above)
[the remark this answers is not in the document]
Comment (clouser2) [reply browser-test/src/admin_api_keys.test.ts]: Works for me.

File: server/app/controllers/admin/AdminApiKeysController.java (hunk 20 above)
Comment (clouser2) [reply server/app/controllers/admin/AdminApiKeysController.java:53]: No worries. Thanks for digging.

File: server/app/views/admin/apikeys/ApiKeyNewOneView.java (hunk 23 above)
Comment (clouser2) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:121]: Still not blocking, but I also might be misunderstanding.

I see that FieldWithLabel.checkbox() sets the input type to checkbox and only sets the "checked" attribute if `setChecked` is called (which it isn't here). The question I have is what setting the value to "true" does here? Can that be removed to avoid confusion?

Comment (bion) [reply server/app/views/admin/apikeys/ApiKeyNewOneView.java:121]: Ah, I've always set checkbox values to "true" out of habit. [Looks like the value is `on` by default](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/checkbox#value) so maybe I've done that over the years because "true" fit better with the boolean concept in my head?

Anyways, thanks for asking! It looks like setting this unnecessary. I'll remove it in my next PR.
